### PR TITLE
Naughton/popover

### PR DESCRIPTION
**Pull request by gusnaughton**
# Alaska Airlines Pull Request

**Fixes:** https://github.com/AlaskaAirlines/auro/issues/22 and https://github.com/AlaskaAirlines/auro/issues/157

## Summary:

Implements auro-popover. 

Pending work:
- [x] Finalize tests for placement validation
- [x] add drop shadow to arrow
- [x] fix karma issue with the process global not existing 

## Type of change:

- [x] New capability

## Checklist:

- [x] My update follows the CONTRIBUTING guidelines of this project
- [x] I have performed a self-review of my own update

**By submitting this Pull Request, I confirm that my contribution is made under the terms of the Apache 2.0 license.**

_Pull Requests will be evaluated by their quality of update and whether it is consistent with the goals and values of this project. Any submission is to be considered a conversation between the submitter and the maintainers of this project and may require changes to your submission._

**Thank you for your submission!**<br>
-- Auro Design System Team

### CODE REVIEW

File: src/auro-popover.js [hunk 1]
```diff
@@ -9,16 +9,20 @@ import { LitElement, html, css } from "lit-element";
...
-  // }
 
-  // function to define props used within the scope of thie component
+  // function to define props used within the scope of this component
   static get properties() {
     return {
-      cssClass:   { type: String }
+      placement:  { type: String },
+      for:        { type: String },
+      tooltip:    { type: Object }
```
Comment (blackfalcon) [inline src/auro-popover.js:25]: Why is this part of the public API. I don't see why/how you would use this at the WC level?

File: src/style.scss [hunk 2]
```diff
@@ -9,24 +9,57 @@
 
 @import "./node_modules/@alaskaairux/orion-web-core-style-sheets/dist/breakpoints";
 @import "./node_modules/@alaskaairux/orion-web-core-style-sheets/dist/core";
+// essentials setup baseline primitive selectors for any UI development
+@import "./node_modules/@alaskaairux/orion-design-tokens/dist/tokens/SCSSVariableMap";
```
Comment (blackfalcon) [inline src/style.scss:13]: I don't think that this is being used anymore. I removed the references in the Sass and the tooltip bubble didn't change? 

Lines 13 - 16

File: src/style.scss [hunk 3]
```diff
@@ -9,24 +9,57 @@
...
+}
+
+.arrow {
+  position: relative;
+  margin-top: -var(--auro-size-xs);
+
+  &:before {
+    content: '';
+    position: absolute;
+    background: white;
```
Comment (blackfalcon) [inline src/style.scss:60]: var(--auro-color-base-white)

File: src/style.scss (hunk 2 above)
Comment (gusnaughton) [reply src/style.scss:13]: Removing this breaks the insetUtility on my machine, are you seeing the same behavior?

File: demo/index.html [hunk 4]
```diff
@@ -13,10 +13,7 @@
       import '@polymer/iron-demo-helpers/demo-snippet';
     </script>
     <script type="module" src="../src/auro-popover.js"></script>
-    <script src="../demo/alert.js"></script>
-    <script type="module">
-      import '../node_modules/@polymer/iron-icons/hardware-icons.js';
-    </script>
+    <script>var process = {env: {}};</script>
```
[the remark this answers is not in the document]
Comment (gusnaughton) [reply demo/index.html:16]: https://github.com/popperjs/popper-core/issues/933#issue-548715007 will do further experimentation

File: src/style.scss (hunk 2 above)
Comment (blackfalcon) [reply src/style.scss]: Removed code. UI not effected. Committed and pushed.

File: demo/index.html (hunk 4 above)
[the remark this answers is not in the document]
Comment (blackfalcon) [reply demo/index.html:16]: So the issue is that we are using popper in a buildless env and the core npm that is to support this, doesn't? So the solution here is to set a process var in the global scope? 

That scares me, a lot.

File: src/auro-popover.js [hunk 5]
```diff
@@ -32,40 +31,56 @@ class AuroPopover extends LitElement {
...
+      popper = createPopper(button, element, {
+        tooltip: element,
         placement: this.placement,
         modifiers: [
         {
           name: 'offset',
           options: {
-            offset: [POPOVER_OFFSET_MIN,POPOVER_OFFSET_MAX],
+            offset: [
+POPOVER_OFFSET_MIN,
```
Comment (blackfalcon) [inline src/auro-popover.js:48]: nitpick: tabbing

**Comment by gusnaughton** [comment]
Issue with tests pending for the event-driven cases:

In the demo, this event is dispatched correctly and the popover fires:
MicrosoftTeams-image (3)

In Karma,[ this event ](https://github.com/AlaskaAirlines/auro-popover/pull/1/files#diff-3b30731663a9559e20c5f1eb02574e3dR29)seems to not dispatch correctly - the same query selector can't be used.
MicrosoftTeams-image (4)

File: README.md [hunk 6]
```diff
@@ -67,16 +67,17 @@ The `polyfills.js` is packaged with this component, but **IT IS NOT NEEDED** to
 
 The `<auro-popover>` element should be used in situations where users may:
 
-* ...
-* ...
-* ...
+* Hover/blur over an element to display a tooltip-style graphic
```
Comment (blackfalcon) [inline README.md:70]: There is a lot of `tool-tip` and `popover` mentions. We need to get tighter on this. For example, in this phrase, I would state this once. 

`Hover/blur over an element to display a tooltip-style popover`

File: README.md [hunk 7]
```diff
@@ -67,16 +67,17 @@ The `polyfills.js` is packaged with this component, but **IT IS NOT NEEDED** to
...
 
 ## API Code Examples
 
 Default auro-popover
 
 ```html
-<auro-popover>Hello World</auro-popover>
+<button id="button1">Click me!</button>
```
Comment (blackfalcon) [inline README.md:77]: Are we saying anywhere in the use docs that you need a corresponding ID for a `for` reference?

File: demo/index.html [hunk 8]
```diff
@@ -27,58 +28,35 @@
...
 
-      <h2>Element &#60;auro-popover&#62;</h2>
-
-      <pre><code>class AuroPopover extends LitElement</code></pre>
-
-      <h3>Styling:</h3>
-      <p>Info ...</p>
+      <demo-snippet>
+        <template>
+            <button id="button2" aria-describedby="tooltip">Popover top</button>
```
Comment (blackfalcon) [inline demo/index.html:40]: You have the `aria-describedby` but there is no corresponding `ID`? 

Also ... for popovers, is `aria-describedby` needed for a11y? If so, we need to have these in the demo docs.

File: demo/index.html [hunk 9]
```diff
@@ -27,58 +28,35 @@
...
+            <auro-popover placement="top" for="button2">Hello World!</auro-popover>
+        </template>
+      </demo-snippet>
 
-      <h3>Responsive state:</h3>
-      <p>Info ...</p>
+      <demo-snippet>
+        <template>
+            <auro-hyperlink id="hyperlink1" href="alaskaair.com" aria-describedby="tooltip">Hover over me!</auro-hyperlink>
```
Comment (blackfalcon) [inline demo/index.html:47]: Again, `aria-describedby="tooltip"` what is this referencing?

File: docs/api.md [hunk 10]
```diff
@@ -1,5 +1,14 @@
...
+# auro-popover
 
-| Attribute | Value type | Description |
-|----|----|----|
-| attr | string, boolean, function | Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. |
+## Properties
+
+| Property    | Attribute   | Type     |
+|-------------|-------------|----------|
+| `for`       | `for`       | `string` |
```
Comment (blackfalcon) [inline docs/api.md:7]: There are no descriptions? See `Describe attributes of your component` at https://auro.alaskaair.com/getting-started/developers/generator/getting-started to add these descriptions.

File: src/auro-popover.js [hunk 11]
```diff
@@ -9,30 +9,89 @@ import { LitElement, html, css } from "lit-element";
 import "focus-visible/dist/focus-visible.min.js";
 import styleCss from "./style-css.js";
 
-// build the component class
-class AuroPopover extends LitElement {
-  // constructor() {
-  //   super();
-  // }
+import Popover from "./tooltip";
```
Comment (blackfalcon) [inline src/auro-popover.js:12]: This is the `popover` and `tooltip` thing. It's weird that there is a `tooltip` js file that pulls in popper.js. 

So, naming is hard. what if you just called this popover.js

File: src/auro-popover.js [hunk 12]
```diff
@@ -9,30 +9,89 @@ import { LitElement, html, css } from "lit-element";
...
 
-  // function to define props used within the scope of thie component
+class AuroPopover extends LitElement {
+  // function to define props used within the scope of this component
   static get properties() {
     return {
-      cssClass:   { type: String }
+      placement:  { type: String },
```
Comment (blackfalcon) [inline src/auro-popover.js:18]: See the previous comment for adding descriptions to these properties.

File: src/auro-popover.js [hunk 13]
```diff
@@ -9,30 +9,89 @@ import { LitElement, html, css } from "lit-element";
...
+    });
+
+  }
+
+
   // function that renders the HTML and CSS into  the scope of the component
   render() {
     return html`
-      <div class=${this.cssClass}>
+      <div id="tooltip" class="tooltip auro util_insetLg" role="tooltip">
```
Comment (blackfalcon) [inline src/auro-popover.js:92]: for the popover there is a `tooltip` CSS class? Again, I think that the naming needs to be more consistent. 

Also, what is the `auro` selector for?

File: src/style.scss [hunk 14]
```diff
@@ -6,27 +6,57 @@
...
+
 
 @import "./node_modules/@alaskaairux/orion-web-core-style-sheets/dist/breakpoints";
 @import "./node_modules/@alaskaairux/orion-web-core-style-sheets/dist/core";
 
-// container styles - defines the look of the outer custom componet
-// :host {
-//   ...
-// }
+.tooltip {
```
Comment (blackfalcon) [inline src/style.scss:16]: there is that name again

File: README.md (hunk 7 above)
Comment (gusnaughton) [reply README.md:79]: Updated in latest

File: demo/index.html (hunk 8 above)
Comment (gusnaughton) [reply demo/index.html:40]: Artifact from a different repository. Removed.

File: docs/api.md [hunk 15]
```diff
@@ -1,5 +1,16 @@
...
-| Attribute | Value type | Description |
-|----|----|----|
-| attr | string, boolean, function | Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. |
+Popover attaches to an element and displays on hover/blur.
+
+## Properties
+
+| Property    | Attribute   | Type     | Description                                      |
+|-------------|-------------|----------|--------------------------------------------------|
+| `for`       | `for`       | `String` | Defines an id for an element in the DOM to trigger on hover/blu. |
```
Comment (blackfalcon) [inline docs/api.md:9]: typo: `hover/blu`

File: src/auro-popover.js [hunk 16]
```diff
@@ -9,30 +9,95 @@ import { LitElement, html, css } from "lit-element";
...
-  //   super();
-  // }
+import Popover from "./popover";
 
-  // function to define props used within the scope of thie component
+/**
+ * Popover attaches to an element and displays on hover/blur.
+ *
+ * @attr {String} placement - Expects top/bottom - default position for popover in relation to the element.
+ * @attr {String} for - Defines an id for an element in the DOM to trigger on hover/blu.
```
Comment (blackfalcon) [inline src/auro-popover.js:18]: typo: `hover/blu`

**Pull request merged**